### ISSUE TITLE
Allow multiple packages to be loaded from a single repo

### ISSUE THREAD

**Issue by kahmali** [DispatchMe/mgp#14]
It is becoming fairly common to have multiple meteor packages stored in a single repository. In favor of keeping the config file as DRY as possible, users should be able to define multiple packages under a single tarball (example in js for comments):

##### Proposal 1
```js
{
  // The typical way of defining a repo with a single Meteor package
  "my:private-package": {
    "tarball": "https://api.github.com/repos/my/private-packages/tarball/commithash",
    "path": "optional/directory/path"
  },
  "my:other-private-package": {
    "tarball": "https://api.github.com/repos/my/private-packages/tarball/commithash"
  },
  // Define a list of packages under a single repo
  "https://api.github.com/repos/my/private-packages/tarball/commithash": [
    {
      "name": "my:first-package-in-repo",
      "path": "path/to/first/package/directory",
    },
    {
      "name": "my:second-package-in-repo",
      "path": "path/to/second/package/directory",
    }
  ],
  "token": "GITHUB_ACCESS_TOKEN"
}
```

Then you could just check to see if each property is an object or array, and when it's an array you just use the same tarball (the JSON key to the array) for each package in the array.

##### Proposal 2
```js
{
  // The typical way of defining a repo with a single Meteor package
  "my:private-package": {
    "tarball": "https://api.github.com/repos/my/private-packages/tarball/commithash",
    "path": "optional/directory/path"
  },
  "my:other-private-package": {
    "tarball": "https://api.github.com/repos/my/private-packages/tarball/commithash"
  },
  // Define a list of packages under a single repo, with some config options
  "https://api.github.com/repos/my/private-packages/tarball/commithash": {
    // All packages will have their paths appended to the basePath
    basePath: "base",
    packages: [
      {
        "name": "my:first-package-in-repo",
        // This package would be in directory base/path/to/first/package/directory
        "path": "path/to/first/package/directory",
      },
      {
        "name": "my:second-package-in-repo",
        "path": "path/to/second/package/directory",
      }
    ]
  },
  "token": "GITHUB_ACCESS_TOKEN"
}
```

In this version, we make things a little bit DRYer, at the cost of a little more work on our end (always a reasonable tradeoff). All top level properties are now objects, so we can't easily differentiate between the two. We would just need to do a schema check on each object in our list to see if the schema matched a package list or individual package definition.

**Comment by creeonix** [comment]
Here is DRY is bringing unnecessary complexity. And to be honest you didn't DRY anything at all because following solves almost everything we need:
```json
{
  "package-name": {
    "repo": "https://github.com/someting/there",
    "tag": "optional-version"
  }
}
```

What I would suggest is to get rid of `"tarball"` and `"token"`, place `"packages.json"` in the root of the repo, read it and extract packages from there. `repo` should show where to download `packages.json`, `tag` should show revision. In `packages.json` we could do following:
```json
{
  "package-name": {
    "path": "some/path/in/the/repo",
    "author": "Or Any Other Meta"
  }
}
```

In this way you will get real repository with packages, and simplify definition of packages to download.

PS: `token` could be optional, but I'd prefer `.netrc` as a default way.

**Comment by kahmali** [comment]
The problem with what you're suggesting is that you're now relying on the package maintainer to also be using mgp, or to be willing to allow you to add a `packages.json` file to their repo. This pattern will be great when you have control over the package repo, but what happens when you don't? What if you only want a subset of the packages within the repo (as we're doing with simple:rest)? Also, how is what I suggested in Proposal 2 not completely DRY?? Please show me one place that I repeated myself. All you're doing is moving the package declaration to a separate file. The complexity is all still there. All you've done is break it across two separate files, which only seems more complex to me. Is the API really that much more difficult to comprehend? Maybe I'm misunderstanding you, but nothing you've suggested seems like an improvement over my second proposal. And I would be concerned about this approach for the reasons I've mentioned here. Again, I'm always willing to add complexity on our end to reduce complexity on a user's end. Allowing them to do everything from their own repo gives them total control, and what I would consider to be a very simple API to use.

**Comment by jperl** [comment]
Proposal 1 and proposal 2 both introduce inconsistency in the definition (sometimes the key is a package name, sometimes a url) which I do not believe is a good pattern. @creeonix I agree we should support .netrc there is an open issue for that #9.

Instead of tarball we should use git:// uris which support commit hash or tag.

**Comment by kahmali** [comment]
@jperl If you don't like the inconsistency then don't support the old, wet way. Just use the tarball url (or git:// uri) as the key as I've suggested. There's no reason someone would use the two different styles in the same file. I was only showing it that way so I could keep it all in one example. Supporting the original style would only be for backwards compatibility (if that's important to you). The new style that I proposed can be used to define a single package in a repo, or multiple. The thing is, there will never be multiple packages with the same name, but there are certainly multiple packages in a single repo. So by categorizing the packages by repo instead of name, you open the possibility for defining multiple packages DRYly. I just don't see what the downside is here. It's all good though. If you're really that against it I'll just fix it in a fork. I need it for my own personal use, and it's a super simple fix.
